libs/@local/hql/diagnostics/src/diagnostic.rs

```diff
@@ -0,0 +1,146 @@
+use core::{
+    error::Error,
+    fmt::{Debug, Display},
+};
+
+use ariadne::ColorGenerator;
+use error_stack::{Report, Result};
+use hql_span::{storage::SpanStorage, tree::SpanNode, Span, SpanId};
+
+use crate::{
+    category::Category,
+    config::ReportConfig,
+    error::ResolveError,
+    help::Help,
+    label::Label,
+    note::Note,
+    severity::Severity,
+    span::{absolute_span, AbsoluteDiagnosticSpan, TransformSpan},
+};
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize))]
+pub struct Diagnostic<S> {
+    pub category: Category,
+    pub severity: Severity,
+
+    pub message: Option<Box<str>>,
+    pub span: Option<S>,
+
+    pub labels: Vec<Label<S>>,
+    pub note: Option<Note>,
+    pub help: Option<Help>,
+}
+
+impl<S> Diagnostic<S> {
+    #[must_use]
+    pub const fn new(category: Category, severity: Severity) -> Self {
+        Self {
+            category,
+            severity,
+            message: None,
+            span: None,
+            labels: Vec::new(),
+            note: None,
+            help: None,
+        }
+    }
+}
+
+impl Diagnostic<SpanId> {
+    /// Resolve the diagnostic, into a proper diagnostic with span nodes.
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if the span id is not found in the span storage.
+    pub fn resolve<S>(
+        self,
+        storage: &SpanStorage<S>,
+    ) -> Result<Diagnostic<SpanNode<S>>, ResolveError>
+    where
+        S: Span + Clone,
+    {
+        let span = self
+            .span
+            .map(|id| {
+                storage
+                    .resolve(id)
+                    .ok_or_else(|| Report::new(ResolveError::UnknownSpan { id }))
+            })
+            .transpose();
+
+        let (span, labels) = self
+            .labels
+            .into_iter()
+            .map(|label| label.resolve(storage))
+            .fold(span.map(|node| (node, Vec::new())), |acc, label| {
+                match (acc, label) {
+                    (Ok((span, mut labels)), Ok(label)) => {
+                        labels.push(label);
+                        Ok((span, labels))
+                    }
+                    (Err(mut acc), Err(error)) => {
+                        acc.extend_one(error);
+                        Err(acc)
+                    }
+                    (Err(acc), _) => Err(acc),
+                    (_, Err(error)) => Err(error),
```

Comment:
Is this the same? Then let's use this:
```suggestion
                    (Err(acc), Ok(_)) => Err(acc),
                    (Ok(_), Err(error)) => Err(error),
```

Reply:
fixed in [475bf01](https://github.com/hashintel/hash/pull/4779/commits/475bf011dd3808731ebd4dfb21625a70917e75fa)